packages/lib/survey/service.ts

```diff
@@ -706,114 +710,188 @@ export const createSurvey = async (
+      productOverwrites: existingSurvey.productOverwrites
+        ? structuredClone(existingSurvey.productOverwrites)
+        : Prisma.JsonNull,
+      styling: existingSurvey.styling ? structuredClone(existingSurvey.styling) : Prisma.JsonNull,
+      segment: undefined,
+    };
 
-    // if the existing survey has an inline segment, we copy the filters and create a new inline segment and connect it to the new survey
+    // Handle segment
     if (existingSurvey.segment) {
       if (existingSurvey.segment.isPrivate) {
-        const newInlineSegment = await createSegment({
-          environmentId,
-          title: `${newSurvey.id}`,
-          isPrivate: true,
-          surveyId: newSurvey.id,
-          filters: existingSurvey.segment.filters,
-        });
-
-        await prisma.survey.update({
-          where: {
-            id: newSurvey.id,
+        surveyData.segment = {
+          create: {
+            title: `Copy of ${existingSurvey.segment.title}`,
+            isPrivate: true,
+            filters: existingSurvey.segment.filters,
+            environment: { connect: { id: targetEnvironmentId } },
           },
-          data: {
-            segment: {
-              connect: {
-                id: newInlineSegment.id,
-              },
-            },
-          },
-        });
-
-        segmentCache.revalidate({
-          id: newInlineSegment.id,
-          environmentId: newSurvey.environmentId,
-        });
+        };
+      } else if (isSameEnvironment) {
+        surveyData.segment = { connect: { id: existingSurvey.segment.id } };
       } else {
-        await prisma.survey.update({
+        const existingSegmentInTargetEnvironment = await prisma.segment.findFirst({
           where: {
-            id: newSurvey.id,
-          },
-          data: {
-            segment: {
-              connect: {
-                id: existingSurvey.segment.id,
-              },
-            },
+            title: existingSurvey.segment.title,
+            isPrivate: false,
+            environmentId: targetEnvironmentId,
           },
         });
 
-        segmentCache.revalidate({
-          id: existingSurvey.segment.id,
-          environmentId: newSurvey.environmentId,
-        });
+        surveyData.segment = {
+          create: {
+            title: existingSegmentInTargetEnvironment
+              ? `${existingSurvey.segment.title}-${Date.now()}`
+              : existingSurvey.segment.title,
+            isPrivate: false,
+            filters: existingSurvey.segment.filters,
+            environment: { connect: { id: targetEnvironmentId } },
+          },
+        };
       }
     }
 
+    // Execute the survey creation in a transaction
+    const newSurvey = await prisma.survey.create({ data: surveyData, select: selectSurvey });
+
+    // Invalidate caches
+    productCache.revalidate({ id: targetProduct.id });
+    actionClassCache.revalidate({ environmentId: targetEnvironmentId });
```

Comment:
We should only invalidate the cache if neccessary since it will create (sometimes expensive) database calls. 
So we should only revalidate the cache, if we change the actionClasses. As I understand the code we also revalidate the cache, when the survey doesn't have an actionClass. Ideally we only revalidate the cache if a new actionClass is created.

Reply:
Also ideally we should also revalidate the other actionClass cache (`byNameAndEnvironmentId`, `byId`) to make sure the cache for actionClasses is correct and reflects the new actionClass correctly.